pkg/controller/direct/securesourcemanager/repository_controller.go

```diff
@@ -0,0 +1,226 @@
+
+	id, err := krm.NewSecureSourceManagerRepositoryRef(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get securesourcemanager GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &SecureSourceManagerRepositoryAdapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *modelSecureSourceManagerRepository) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type SecureSourceManagerRepositoryAdapter struct {
+	id        *krm.SecureSourceManagerRepositoryRef
+	gcpClient *gcp.Client
+	desired   *krm.SecureSourceManagerRepository
+	actual    *securesourcemanagerpb.Repository
+}
+
+var _ directbase.Adapter = &SecureSourceManagerRepositoryAdapter{}
+
+func (a *SecureSourceManagerRepositoryAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting SecureSourceManagerRepository", "name", a.id.External)
+
+	req := &securesourcemanagerpb.GetRepositoryRequest{Name: a.id.External}
+	repositorypb, err := a.gcpClient.GetRepository(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting SecureSourceManagerRepository %q: %w", a.id.External, err)
+	}
+
+	a.actual = repositorypb
+	return true, nil
+}
+
+func (a *SecureSourceManagerRepositoryAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating Repository", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := SecureSourceManagerRepositorySpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	parent, err := a.id.Parent()
+	if err != nil {
+		return err
+	}
+	repositoryID, err := a.id.ResourceID()
+	if err != nil {
+		return err
+	}
+
+	req := &securesourcemanagerpb.CreateRepositoryRequest{
+		Parent:       parent.String(),
+		Repository:   resource,
+		RepositoryId: repositoryID,
+	}
+	op, err := a.gcpClient.CreateRepository(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Repository %s: %w", a.id.External, err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("Repository %s waiting creation: %w", a.id.External, err)
+	}
+	log.V(2).Info("successfully created Repository", "name", a.id.External)
+
+	status := &krm.SecureSourceManagerRepositoryStatus{}
+	status.ObservedState = SecureSourceManagerRepositoryObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = &a.id.External
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *SecureSourceManagerRepositoryAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.Info("update of SecureSourceManagerRepository not supported")
+	return nil
+}
```

Comment:
Want to double check on this. Are `instance` and `instanceConfig` fields immutable? Can they be updated in the future?

Reply:
There's no `instanceConfig` field. (Maybe you thought of the `initialConfig` field?)

For the `instance` field, no near-term plans to make it updatable. The `initialConfig` field is an input only field.

Reply:
shall we mark `instance` and `initialConfig` immutable [here](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/apis/securesourcemanager/v1alpha1/repository_types.go#L43-L49)?

Reply:
Discussed offline, let's follow up on this in another PR!

Reply:
Will do!